### DISCUSSION TITLE
Templating of "service_data" ?

### DISCUSSION THREAD

**Discussion by misko2k** [ExperienceLovelace/ha-floorplan#48] [Q&A] [answered]
Now i am configuring my floorplans , and there is something what i am missing here , templating of conditions for example  : 
I use often complex conditions in "service_data" (for floorplan.text_set and floorplan.class_set, and i must every time go and change an X places my code , instead to write 1 code stored in text file what i can include in needed position , is this possible in HA-floorplan ? 

![templating](https://user-images.githubusercontent.com/21138890/107264676-8c0dcd80-6a43-11eb-8de3-d8e2ee4d1b2e.jpg)

**Comment by pkozul** [comment]
Hi @misko2k 

Floorplan currently does not have support for loading external code files, or even loading 'common' functions that are external to the rules themselves.

I think it would make sense to add support for 'common' functions, which could be defined in a new section of the floorplan config. We will consider this.

**Comment by pkozul** [comment, marked as answer]
@misko2k 

This was actually quite easy to implement. New version 1.0.17 has support for functions. You can see this in the Ring example.

https://experiencelovelace.github.io/ha-floorplan/docs/example-ring/

There is a new (optional) config section called `functions:`. You can add as many functions as you need. They just need to be contained in the structure shown:

```
  functions: |
    >
    return {

      getPercentageFill: (entity) => {
        var max = [195, 232, 141];
        var min = [240, 113, 120];
        var r = Math.floor(min[0] + ((max[0] - min[0]) * (entity.state / 100)));
        var g = Math.floor(min[1] + ((max[1] - min[1]) * (entity.state / 100)));
        var b = Math.floor(min[2] + ((max[2] - min[2]) * (entity.state / 100)));
        return `fill: rgb(${r}, ${g}, ${b})`; 
      },

      someOtherFunctionA: (entity, entities, hass) => {
        return 'foo'; 
      },

      someOtherFunctionB: (entity, entities, hass) => {
        return 'bar'; 
      },
      
    };
```

Then in your rules, you have access to all your functions, which are accessible via the `functions` object in your code (i.e. `functions.getPercentageFill()`):

```
    - entity: sensor.ring_salon_battery
      state_action:
        - service: floorplan.style_set
          service_data: ${functions.getPercentageFill(entity)}
```

Try that and let us know how it goes.

**Reply by exetico** [reply]
It's now released :-)
